File: node/src/py/message_handler.py
import uuid
from asyncio import StreamReader, StreamWriter
from datetime import datetime, timezone
import node.src.py.util as util
from lib.src.types.py.block import Block, BlockHeader
from lib.src.types.py.crypto import Hash, MerkleRoot
from lib.src.types.py.transaction import Transaction, TransactionOutput
from lib.src.types.py.network import (
    UTXOs, Template, Difference, TemplateValidity, NodeList,
    FetchBlock, DiscoverNodes, AskDifference, FetchUTXOs, NewBlock,
    NewTransaction, ValidateTemplate, SubmitTemplate, SubmitTransaction,
    FetchTemplate, send_message, receive_message
)

# Maximum of transactions allowed in a block
BLOCK_TRANSACTION_CAP = 20

async def handle(reader: StreamReader, writer: StreamWriter):
    try:
        while True:
            # Read a message from the socket
            try:
                message = await receive_message(reader)
            except Exception as e:
                print(f"Invalid message from peer: {e}, closing connection")
                return

            match message:

                case UTXOs() | Template() | Difference() | TemplateValidity() | NodeList():
                    print("I am neither a miner nor a wallet! Goodbye")
                    writer.close()
                    await writer.wait_closed()
                    return

                case FetchBlock(height=height):
                    blockchain = util.BLOCKCHAIN
                    try:
                        block = blockchain.blocks[height]
                        response = NewBlock(block=block)
                        await send_message(writer, response)
                    except IndexError:
                        return

                case DiscoverNodes():
                    nodes = list(util.NODES.keys())
                    response = NodeList(nodes=nodes)
                    await send_message(writer, response)

                case AskDifference(height=height):
                    blockchain = util.BLOCKCHAIN
                    count = len(blockchain.blocks) - height
                    response = Difference(diff=count)
                    await send_message(writer, response)

                case FetchUTXOs(public_key=public_key):
                    print("Received request to fetch UTXOs")
                    blockchain = util.BLOCKCHAIN
                    utxos = []
                    for _, (marked, tx_out) in blockchain.utxos.items():
                        if tx_out.public_key == public_key:
                            utxos.append((tx_out, marked))
                    response = UTXOs(utxos=utxos)
                    await send_message(writer, response)

                case NewBlock(block=block):
                    blockchain = util.BLOCKCHAIN
                    print("Received new block")
                    try:
                        blockchain.add_block(block)
                    except Exception as e:
                        print(f"Block rejected: {e}")

                case NewTransaction(transaction=transaction):
                    blockchain = util.BLOCKCHAIN
                    print("Received transaction from friend")
                    try:
                        blockchain.add_to_mempool(transaction)
                    except Exception as e:
                        print(f"Transaction rejected, closing connection: {e}")
                        writer.close()
                        await writer.wait_closed()
                        return

                case ValidateTemplate(block=block):
                    blockchain = util.BLOCKCHAIN
                    last_block_hash = blockchain.blocks[-1].hash() if blockchain.blocks else Hash.zero()
                    status = block.header.prev_hash == last_block_hash
                    response = TemplateValidity(valid=status)
                    await send_message(writer, response)

                case SubmitTemplate(block=block):
                    print("Received allegedly mined template")
                    blockchain = util.BLOCKCHAIN
                    try:
                        blockchain.add_block(block)
                    except Exception as e:
                        print(f"Block rejected: {e}, closing connection")
                        writer.close()
                        await writer.wait_closed()
                        return
                    blockchain.rebuild_utxos()
                    print("Block looks good, broadcasting")
                    
                    # Send block to all friend nodes
                    nodes = list(util.NODES.keys())
                    for node in nodes:
                        _, node_writer = util.NODES[node]
                        response = NewBlock(block=block)
                        try:
                            await send_message(node_writer, response)
                        except Exception as e:
                            print(f"failed to send block to {node}: {e}")

                case SubmitTransaction(transaction=transaction):
                    print("Submit tx")
                    blockchain = util.BLOCKCHAIN
                    try:
                        blockchain.add_to_mempool(transaction)
                    except Exception as e:
                        print(f"transaction rejected, closing connection: {e}")
                        writer.close()
                        await writer.wait_closed()
                        return
                    
                    print("Added transaction to mempool")
                    # Send transaction to all friend nodes
                    nodes = list(util.NODES.keys())
                    for node in nodes:
                        print(f"sending to friend: {node}")
                        _, node_writer = util.NODES[node]
                        response = NewTransaction(transaction=transaction)
                        try:
                            await send_message(node_writer, response)
                        except Exception as e:
                            print(f"failed to send transaction to {node}: {e}")
                    print("transaction sent to friends")

                case FetchTemplate(public_key=public_key):
                    blockchain = util.BLOCKCHAIN
                    transactions = []
                    # Insert transactions from mempool
                    transactions.extend([tx for _, tx in blockchain.mempool[:BLOCK_TRANSACTION_CAP]])
                    # Insert coinbase tx with the public key
                    coinbase_tx = Transaction(
                        inputs=[],
                        outputs=[TransactionOutput(
                            public_key=public_key,
                            unique_id=uuid.uuid4(),
                            value=0
                        )]
                    )
                    transactions.insert(0, coinbase_tx)
                    merkle_root = MerkleRoot.calculate(transactions)
                    last_block_hash = blockchain.blocks[-1].hash() if blockchain.blocks else Hash.zero()
                    block = Block(
                        header=BlockHeader(
                            timestamp=datetime.now(timezone.utc),
                            prev_hash=last_block_hash,
                            nonce=0,
                            target=blockchain.target,
                            merkle_root=merkle_root
                        ),
                        transactions=transactions
                    )
                    try:
                        miner_fees = block.calculate_miner_fees(blockchain.utxos)
                    except Exception as e:
                        print(f"{e}")
                        return
                    # Reward calculation: 50 * 10^8 / 2^(height / 210)
                    predicted_height = len(blockchain.blocks)
                    reward = (50 * 10**8) // (2**(predicted_height // 210))
                    # Update coinbase tx with reward
                    block.transactions[0].outputs[0].value = reward + miner_fees
                    # Recalculate merkle root
                    block.header.merkle_root = MerkleRoot.calculate(block.transactions)
                    response = Template(block=block)
                    await send_message(writer, response)
    finally:
        writer.close()
        await writer.wait_closed()

File: node/src/message_handler.rs
use std::collections::HashSet;
use std::sync::Arc;
use base64::Engine;
use base64::engine::general_purpose;
use chrono::Utc;
use log::{error, info, warn};
use uuid::Uuid;
use tokio::net::TcpStream;
use tokio::sync::Mutex;
use btclib::crypto::{Hash, MerkleRoot};
use btclib::network::Message;
use btclib::types::{Block, BlockHeader, Transaction, TransactionOutput};
use btclib::network::Message::*;

pub async fn handle(stream: TcpStream) {
    let stream = Arc::new(Mutex::new(stream));
    loop {
        // Read a message from the socket
        let mut locked_stream = stream.lock().await;
        let message = match Message::receive_async(&mut *locked_stream).await {
            Ok(message) => message,
            Err(e) => {
                error!("Invalid message from peer: {e}, closing connection");
                return;
            }
        };
        match message {
            AskDifference(height) => {
                let blockchain = crate::BLOCKCHAIN.read().await;
                let count = blockchain.block_height() as i32 - height as i32;
                let message = Difference(count);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
            DiscoverNodes(dialing_node, current_node) => {
                // Here, the responding node is the dialed one
                info!("📞 [{}] receiving call from [{}]", current_node, dialing_node);
                match TcpStream::connect(&dialing_node).await {
                    Ok(s) => {
                        let stream = Arc::new(Mutex::new(s));
                        crate::NODES.insert(dialing_node.clone(), stream);
                        info!("➕  Added node [{}]", dialing_node);
                        info!("🌐 Known network nodes: [{}]", crate::NODES.len());
                    },
                    Err(e) => {
                        error!("⚠️ Failed to connect to {}: {}", dialing_node, e);
                        return;
                    }
                }
                let nodes: HashSet<String> = crate::NODES.iter().map(|x| x.key().clone()).collect();
                let message = NodeList(nodes);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
            FetchBlock(height) => {
                let blockchain = crate::BLOCKCHAIN.read().await;
                let Some(block) = blockchain.blocks().nth(height).cloned() else {
                    return;
                };
                let message = NewBlock(block);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
            FetchTemplate(pubkey) => {
                let blockchain = crate::BLOCKCHAIN.read().await;
                let mut transactions = vec![];
                // Insert transactions from mempool
                transactions.extend(
                    blockchain
                        .mempool()
                        .iter()
                        .take(btclib::BLOCK_TRANSACTION_CAP)
                        .map(|(_, tx)| tx)
                        .cloned()
                        .collect::<Vec<_>>(),
                );
                // Insert coinbase tx with a pubkey
                transactions.insert(0, Transaction {
                    inputs: vec![],
                    outputs: vec![TransactionOutput {
                        public_key: pubkey,
                        unique_id: Uuid::new_v4(),
                        value: 0,
                    }]
                });
                let merkle_root = MerkleRoot::calculate(&transactions);
                let mut block = Block::new(
                    BlockHeader {
                        timestamp: Utc::now(),
                        prev_block_hash: blockchain.blocks().last().map(|last_block| {
                            last_block.hash()
                        }).unwrap_or(Hash::zero()),
                        nonce: 0,
                        target: blockchain.target(),
                        merkle_root
                    },
                    transactions
                );
                let miner_fees = match block.calculate_miner_fees(blockchain.utxos()) {
                    Ok(fees) => fees,
                    Err(e) => {
                        error!("{e}");
                        return;
                    }
                };
                let reward =  blockchain.calculate_block_reward();
                // Update coinbase tx with reward
                block.transactions[0].outputs[0].value = reward + miner_fees;
                // Recalculate merkle root
                block.header.merkle_root = MerkleRoot::calculate(&block.transactions);
                let message = Template(block);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
            FetchUTXOs(key) => {
                println!("Received request to fetch UTXOs");
                let blockchain = crate::BLOCKCHAIN.read().await;
                let utxos = blockchain.utxos().iter().filter(|(_, (_, tx_out))| {
                    tx_out.public_key == key
                }).map(|(_, (marked, tx_out))| {
                    (tx_out.clone(), *marked)
                }).collect::<Vec<_>>();
                let message = UTXOs(utxos);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
            NewBlock(block) => {
                let mut blockchain = crate::BLOCKCHAIN.write().await;
                info!("📦 Received new block");
                if blockchain.add_block(block).is_err() {
                    error!("❌  Block rejected");
                }
            }
            NewTransaction(tx) => {
                let mut blockchain = crate::BLOCKCHAIN.write().await;
                println!("Received transaction from friend");
                if blockchain.add_to_mempool(tx).is_err() {
                    error!("❌ Transaction rejected, closing connection");
                    return;
                }
            }
            SubmitTemplate(block, miner) => {
                let encoded_point = miner.0.to_encoded_point(true);
                let miner_id = general_purpose::STANDARD.encode(encoded_point.as_bytes());
                info!("Received allegedly mined template from: 👷{}", miner_id);
                let mut blockchain = crate::BLOCKCHAIN.write().await;
                if let Err(e) = blockchain.add_block(block.clone()) {
                    error!("❌  Block rejected: {e}, closing connection");
                    return;
                }
                blockchain.rebuild_utxos();
                info!("Block looks good, broadcasting📡️");
                // Send block to all friend nodes
                let nodes = crate::NODES.iter().map(|x| x.key().clone()).collect::<Vec<_>>();
                for node in nodes {
                    if let Some(stream) = crate::NODES.get_mut(&node) {
                        let message = NewBlock(block.clone());
                        let mut locked_stream = stream.lock().await;
                        if message.send_async(&mut *locked_stream).await.is_err() {
                            error!("⚠️ Failed to send block to {}", node);
                        }
                    }
                }
            }
            SubmitTransaction(tx) => {
                println!("Submit tx");
                let mut blockchain = crate::BLOCKCHAIN.write().await;
                if let Err(e) = blockchain.add_to_mempool(tx.clone()) {
                    error!("❌ Transaction rejected, closing connection: {e}");
                    return;
                }
                info!("🗃️ Added transaction to mempool");
                // Send transaction to all friend nodes
                let nodes = crate::NODES
                    .iter()
                    .map(|x| x.key().clone())
                    .collect::<Vec<_>>();
                for node in nodes {
                    info!("Sending to friend: [{node}]");
                    if let Some(stream) = crate::NODES.get_mut(&node) {
                        let message = NewTransaction(tx.clone());
                        let mut locked_stream = stream.lock().await;
                        if message.send_async(&mut *locked_stream).await.is_err() {
                            error!("⚠️ Failed to send transaction to {}", node);
                        }
                    }
                }
                info!("💰 Transaction sent to friends");
            }
            UTXOs(_) | Template(_) | Difference(_) | TemplateValidity(_) | NodeList(_) => {
                warn!("👋 I am neither a miner nor a wallet! Goodbye");
                return;
            }
            ValidateTemplate(block_template) => {
                let blockchain = crate::BLOCKCHAIN.read().await;
                let status = block_template.header.prev_block_hash == blockchain
                    .blocks()
                    .last()
                    .map(|last_block| last_block.hash())
                    .unwrap_or(Hash::zero());
                let message = TemplateValidity(status);
                message.send_async(&mut *locked_stream).await.unwrap();
            }
        }
    }
}